astrocut/cube_cut.py

```diff
@@ -800,3 +810,183 @@ def cube_cut(self, cube_file, coordinates, cutout_size,
+
+
+class S3CubeFile():
+    """Interface for S3-hosted AstroCut cube files.
+
+    Inspired by earlier proto-types written by Thomas Robitaille, P. L. Lim,
+    Susan Mullally, Joseph Curtin, and others.
+    """
+
+    # Byte position of the start of HDU[1]'s header
+    HDU1_HEADER_OFFSET = 2880
+    # Byte position of the start of HDU[1]'s data
+    HDU1_DATA_OFFSET = 5760
+
+    def __init__(self, cube_file):
+        if not cube_file.startswith("s3://"):
+            raise ValueError("S3CubeFile expects a cube_file with prefix 's3://'")
+
+        self.cube_file = cube_file
+        match = re.findall("s3://([^/]*)/(.*)", cube_file)[0]
+        self.s3_bucket, self.s3_key = match[0], match[1]
+
+        # The data type is hard-coded to be float32 for TessCut cubes
+        data_type = np.float32
+        self.data_type = np.dtype(data_type).newbyteorder('>')
+        self.itemsize = np.dtype(data_type).itemsize
+
+        # Setup asyncio
+        # asyncio cannot be used in a Jupyter notebook environment
+        # without first calling `nest_asyncio.apply()` following:
+        import nest_asyncio
+        nest_asyncio.apply()
+
+        # Setup S3 client
+        import aioboto3
+        from botocore import UNSIGNED
+        from botocore.config import Config
+        self.s3clientmgr = aioboto3.client("s3", config=Config(signature_version=UNSIGNED))
+        self.s3_client = asyncio.run(self.s3clientmgr.__aenter__())
+
+        # Read the headers of HDU0 and HDU1
+        self.primary_header, self.header = self._read_headers()
+
+        # The shape is typically equal to (2078, 2136, n_cadences, 2)
+        # and can be interpreted as (n_rows, n_columns, n_times, n_flux_types),
+        # matching the FITS header values (NAXIS4, NAXIS3, NAXIS2, NAXIS1).
+        # Caveat: FITS and Python use reversed shape notations!
+        self.shape = tuple([self.header['NAXIS{0}'.format(i + 1)]
+                            for i in range(self.header['NAXIS'])][::-1])
+
+        # strides detail how many bytes separate elements along different dimensions
+        self.strides = (
+            self.shape[1] * self.shape[2] * self.shape[3] * self.itemsize,
+            self.shape[2] * self.shape[3] * self.itemsize,
+            self.shape[3] * self.itemsize,
+            self.itemsize)
+
+    def __enter__(self):
+        return self
+
+    async def __aenter__(self):
+        return self
+
+    def __exit__(self, *args):
+        asyncio.run(self.__aexit__(*args))
+
+    async def __aexit__(self, exc_type, exc_value, exc_traceback):
+        await self.s3clientmgr.__aexit__(exc_type, exc_value, exc_traceback)
+
+    @property
+    def table(self):
+        return self._read_table()
+
+    @lru_cache()
+    def _read_table(self):
+        """Returns the BinTableHDU for HDU #2."""
+        hdu1_data_size = np.product(self.shape) * self.itemsize
+        hdu2_offset = self.HDU1_DATA_OFFSET + 2880 * (1 + hdu1_data_size // 2880)
```

Comment:
Magic number 2880, should this be HDU1_HEADER_OFFSET?

Reply:
2880 here refers to the fact that the FITS standard prescribed that *"each FITS structure shall consist of an integral number of FITS blocks, which are each 2880 bytes"*.  This deserves a comment in the code at minimum, and of course really ought to be defined as a `FITS_BLOCK_SIZE` constant instead of being hard-coded.